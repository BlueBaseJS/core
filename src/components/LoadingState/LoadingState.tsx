import { ActivityIndicator, ComponentState } from '../../index';
import { BlueBaseContext } from '../../Context';
import { ComponentStateProps } from '../ComponentState';
import React from 'react';

export interface LoadingStateProps {

	/** Flag if loading has timedOut. */
	timedOut?: boolean,

	/** Callback function when Retry button is pressed. */
	retry?: () => void,
}

/**
 * # ‚è≥ LoadingState
 *
 * A component that is used to show a loading state. Shows a spinner by
 * default. If 'timedOut' flag is set then it shows a timeout version.
 *
 * ## Usage
 * ```jsx
 * <LoadingState timedOut={false} retry={retryFunction}/>
 * ```
 *
 * TODO: Add a prop to allow custom text for retry button.
 */
export class LoadingState extends React.Component<LoadingStateProps> {

	static contextType = BlueBaseContext;

	render() {

		const { timedOut, retry } = this.props;

		const props: ComponentStateProps = {
			image: <ActivityIndicator />
		};

		if (timedOut === true) {
			props.description = 'This is taking longer than usual';

			if (retry) {
				props.actionTitle = 'Retry';
				props.actionOnPress = retry;
			}
		}

		return <ComponentState {...props} />;
	}
}
