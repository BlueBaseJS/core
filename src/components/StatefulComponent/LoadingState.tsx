import { BlueRain, withBlueRain } from '../../index';
import React from 'react';

const LoadingState = (props: { bluerain: BlueRain }) => {

	const BR = props.bluerain;

	const Indicator = () => <BR.Components.ActivityIndicator color="#eeeeee" size="large" />;

	return <BR.Components.ComponentState title="Loading, Please wait." image={Indicator} />;
};

export default withBlueRain(LoadingState);
