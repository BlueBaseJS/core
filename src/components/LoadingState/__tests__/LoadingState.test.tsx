import { BlueBaseApp } from '../../../';
import { LoadingState } from '../LoadingState';
import React from 'react';
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';


describe('LoadingState', () => {

	test(`should render an ActivityIndicator`, async () => {
		const component = mount(
			<BlueBaseApp>
				<LoadingState />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, LoadingState);
		// expect(component).toMatchSnapshot();

		expect(component.exists('LoadingState ActivityIndicator')).toBe(true);
		expect(component.exists('LoadingState Body2')).toBe(false);
		expect(component.exists('LoadingState Button')).toBe(false);

	});

	test(`should render timeout message`, async () => {
		const component = mount(
			<BlueBaseApp>
				<LoadingState timedOut={true} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, LoadingState);
		// expect(component).toMatchSnapshot();

		expect(component.exists('LoadingState ActivityIndicator')).toBe(true);
		expect(component.exists('LoadingState Body2')).toBe(true);
		expect(component.find('LoadingState Body2 Text').last().text()).toBe('This is taking longer than usual');
		expect(component.exists('LoadingState Button')).toBe(false);
	});

	test(`should render timeout message & retry button`, async () => {
		const mockedFn = jest.fn();

		const component = mount(
			<BlueBaseApp>
				<LoadingState timedOut={true} retry={mockedFn} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, LoadingState);
		// expect(component).toMatchSnapshot();

		expect(component.exists('LoadingState ActivityIndicator')).toBe(true);
		expect(component.exists('LoadingState Body2')).toBe(true);
		expect(component.find('LoadingState Body2 Text').last().text()).toBe('This is taking longer than usual');
		expect(component.exists('LoadingState Button')).toBe(true);

		(component as any).find('LoadingState Button').first().props().onPress();
		expect(mockedFn.mock.calls.length).toBe(1);
	});

});
