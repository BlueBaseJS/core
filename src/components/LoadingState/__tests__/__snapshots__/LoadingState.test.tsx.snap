// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingState should render an ActivityIndicator 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <SystemApp>
        <ThemedComponent>
          <SystemApp>
            <LoadingState>
              <ComponentState
                image={<ActivityIndicator />}
              >
                <ThemedComponent
                  image={<ActivityIndicator />}
                >
                  <ComponentState
                    image={<ActivityIndicator />}
                    styles={
                      Object {
                        "actionRoot": Object {
                          "marginTop": 8,
                        },
                        "description": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                        "image": Object {
                          "height": 250,
                          "width": 250,
                        },
                        "imageRoot": Object {
                          "marginBottom": 16,
                        },
                        "root": Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        },
                        "title": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        }
                      }
                    >
                      <ThemedComponent
                        style={
                          Object {
                            "alignItems": "center",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "justifyContent": "center",
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                                "justifyContent": "center",
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginBottom": 16,
                                }
                              }
                            >
                              <ThemedComponent
                                style={
                                  Object {
                                    "marginBottom": 16,
                                  }
                                }
                              >
                                <View
                                  style={
                                    Object {
                                      "marginBottom": 16,
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "marginBottom": 16,
                                      }
                                    }
                                  >
                                    <ActivityIndicator>
                                      <ThemedComponent>
                                        <ActivityIndicator
                                          animating={true}
                                          color="#999999"
                                          hidesWhenStopped={true}
                                          size="small"
                                        >
                                          <ActivityIndicator
                                            animating={true}
                                            color="#999999"
                                            hidesWhenStopped={true}
                                            size="small"
                                          />
                                        </ActivityIndicator>
                                      </ThemedComponent>
                                    </ActivityIndicator>
                                  </View>
                                </View>
                              </ThemedComponent>
                            </View>
                          </View>
                        </View>
                      </ThemedComponent>
                    </View>
                  </ComponentState>
                </ThemedComponent>
              </ComponentState>
            </LoadingState>
          </SystemApp>
        </ThemedComponent>
      </SystemApp>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`LoadingState should render timeout message & retry button 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <SystemApp>
        <ThemedComponent>
          <SystemApp>
            <LoadingState
              retry={[MockFunction]}
              timedOut={true}
            >
              <ComponentState
                actionOnPress={[MockFunction]}
                actionTitle="Retry"
                description="This is taking longer than usual"
                image={<ActivityIndicator />}
              >
                <ThemedComponent
                  actionOnPress={[MockFunction]}
                  actionTitle="Retry"
                  description="This is taking longer than usual"
                  image={<ActivityIndicator />}
                >
                  <ComponentState
                    actionOnPress={[MockFunction]}
                    actionTitle="Retry"
                    description="This is taking longer than usual"
                    image={<ActivityIndicator />}
                    styles={
                      Object {
                        "actionRoot": Object {
                          "marginTop": 8,
                        },
                        "description": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                        "image": Object {
                          "height": 250,
                          "width": 250,
                        },
                        "imageRoot": Object {
                          "marginBottom": 16,
                        },
                        "root": Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        },
                        "title": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        }
                      }
                    >
                      <ThemedComponent
                        style={
                          Object {
                            "alignItems": "center",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "justifyContent": "center",
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                                "justifyContent": "center",
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginBottom": 16,
                                }
                              }
                            >
                              <ThemedComponent
                                style={
                                  Object {
                                    "marginBottom": 16,
                                  }
                                }
                              >
                                <View
                                  style={
                                    Object {
                                      "marginBottom": 16,
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "marginBottom": 16,
                                      }
                                    }
                                  >
                                    <ActivityIndicator>
                                      <ThemedComponent>
                                        <ActivityIndicator
                                          animating={true}
                                          color="#999999"
                                          hidesWhenStopped={true}
                                          size="small"
                                        >
                                          <ActivityIndicator
                                            animating={true}
                                            color="#999999"
                                            hidesWhenStopped={true}
                                            size="small"
                                          />
                                        </ActivityIndicator>
                                      </ThemedComponent>
                                    </ActivityIndicator>
                                  </View>
                                </View>
                              </ThemedComponent>
                            </View>
                            <Body2
                              style={
                                Object {
                                  "marginBottom": 8,
                                  "textAlign": "center",
                                }
                              }
                            >
                              <ThemedComponent
                                style={
                                  Object {
                                    "marginBottom": 8,
                                    "textAlign": "center",
                                  }
                                }
                              >
                                <Component
                                  style={
                                    Object {
                                      "marginBottom": 8,
                                      "textAlign": "center",
                                    }
                                  }
                                  styles={
                                    Object {
                                      "root": Object {
                                        "color": "rgba(0, 0, 0, 0.87)",
                                        "fontFamily": "Roboto",
                                        "fontSize": 14,
                                        "fontWeight": "400",
                                      },
                                    }
                                  }
                                >
                                  <Text
                                    style={
                                      Array [
                                        Object {
                                          "color": "rgba(0, 0, 0, 0.87)",
                                          "fontFamily": "Roboto",
                                          "fontSize": 14,
                                          "fontWeight": "400",
                                        },
                                        Object {
                                          "marginBottom": 8,
                                          "textAlign": "center",
                                        },
                                      ]
                                    }
                                  >
                                    <Text
                                      style={
                                        Array [
                                          Object {
                                            "color": "rgba(0, 0, 0, 0.87)",
                                            "fontFamily": "Roboto",
                                            "fontSize": 14,
                                            "fontWeight": "400",
                                          },
                                          Object {
                                            "marginBottom": 8,
                                            "textAlign": "center",
                                          },
                                        ]
                                      }
                                    >
                                      This is taking longer than usual
                                    </Text>
                                  </Text>
                                </Component>
                              </ThemedComponent>
                            </Body2>
                            <View
                              style={
                                Object {
                                  "marginTop": 8,
                                }
                              }
                            >
                              <ThemedComponent
                                style={
                                  Object {
                                    "marginTop": 8,
                                  }
                                }
                              >
                                <View
                                  style={
                                    Object {
                                      "marginTop": 8,
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "marginTop": 8,
                                      }
                                    }
                                  >
                                    <Button
                                      color="primary"
                                      onPress={[MockFunction]}
                                    >
                                      <ThemedComponent
                                        color="primary"
                                        onPress={[MockFunction]}
                                      >
                                        <Button
                                          color="primary"
                                          onPress={[MockFunction]}
                                          styles={
                                            Object {
                                              "default": Object {
                                                "backgroundColor": "#fafafa",
                                              },
                                              "defaultText": Object {},
                                              "fullWidth": Object {
                                                "alignSelf": "stretch",
                                              },
                                              "link": Object {},
                                              "linkText": Object {},
                                              "primary": Object {
                                                "backgroundColor": "#3f51b5",
                                              },
                                              "primaryText": Object {
                                                "color": "#fff",
                                              },
                                              "root": Object {
                                                "alignSelf": "flex-start",
                                                "borderRadius": 4,
                                                "paddingBottom": 8,
                                                "paddingLeft": 16,
                                                "paddingRight": 16,
                                                "paddingTop": 8,
                                              },
                                              "secondary": Object {
                                                "backgroundColor": "#f50057",
                                              },
                                              "secondaryText": Object {
                                                "color": "#fff",
                                              },
                                              "text": Object {
                                                "color": "rgba(0, 0, 0, 0.87)",
                                                "fontFamily": "Roboto",
                                                "fontSize": 14,
                                                "fontWeight": "500",
                                                "textAlign": "center",
                                                "textTransform": "uppercase",
                                              },
                                            }
                                          }
                                        >
                                          <TouchableOpacity
                                            accessibilityRole="button"
                                            activeOpacity={0.2}
                                            onPress={[MockFunction]}
                                          >
                                            <AnimatedComponent
                                              accessibilityRole="button"
                                              accessible={true}
                                              isTVSelectable={true}
                                              onResponderGrant={[Function]}
                                              onResponderMove={[Function]}
                                              onResponderRelease={[Function]}
                                              onResponderTerminate={[Function]}
                                              onResponderTerminationRequest={[Function]}
                                              onStartShouldSetResponder={[Function]}
                                              style={
                                                Array [
                                                  undefined,
                                                  Object {
                                                    "opacity": 1,
                                                  },
                                                ]
                                              }
                                            >
                                              <View
                                                accessibilityRole="button"
                                                accessible={true}
                                                isTVSelectable={true}
                                                onResponderGrant={[Function]}
                                                onResponderMove={[Function]}
                                                onResponderRelease={[Function]}
                                                onResponderTerminate={[Function]}
                                                onResponderTerminationRequest={[Function]}
                                                onStartShouldSetResponder={[Function]}
                                                style={
                                                  Object {
                                                    "opacity": 1,
                                                  }
                                                }
                                              >
                                                <View
                                                  accessibilityRole="button"
                                                  accessible={true}
                                                  isTVSelectable={true}
                                                  onResponderGrant={[Function]}
                                                  onResponderMove={[Function]}
                                                  onResponderRelease={[Function]}
                                                  onResponderTerminate={[Function]}
                                                  onResponderTerminationRequest={[Function]}
                                                  onStartShouldSetResponder={[Function]}
                                                  style={
                                                    Object {
                                                      "opacity": 1,
                                                    }
                                                  }
                                                >
                                                  <View
                                                    style={
                                                      Array [
                                                        Object {
                                                          "alignSelf": "flex-start",
                                                          "borderRadius": 4,
                                                          "paddingBottom": 8,
                                                          "paddingLeft": 16,
                                                          "paddingRight": 16,
                                                          "paddingTop": 8,
                                                        },
                                                        Object {
                                                          "backgroundColor": "#3f51b5",
                                                        },
                                                      ]
                                                    }
                                                  >
                                                    <View
                                                      style={
                                                        Array [
                                                          Object {
                                                            "alignSelf": "flex-start",
                                                            "borderRadius": 4,
                                                            "paddingBottom": 8,
                                                            "paddingLeft": 16,
                                                            "paddingRight": 16,
                                                            "paddingTop": 8,
                                                          },
                                                          Object {
                                                            "backgroundColor": "#3f51b5",
                                                          },
                                                        ]
                                                      }
                                                    >
                                                      <Text
                                                        style={
                                                          Array [
                                                            Object {
                                                              "color": "rgba(0, 0, 0, 0.87)",
                                                              "fontFamily": "Roboto",
                                                              "fontSize": 14,
                                                              "fontWeight": "500",
                                                              "textAlign": "center",
                                                              "textTransform": "uppercase",
                                                            },
                                                            Object {
                                                              "color": "#fff",
                                                            },
                                                          ]
                                                        }
                                                      >
                                                        <Text
                                                          style={
                                                            Array [
                                                              Object {
                                                                "color": "rgba(0, 0, 0, 0.87)",
                                                                "fontFamily": "Roboto",
                                                                "fontSize": 14,
                                                                "fontWeight": "500",
                                                                "textAlign": "center",
                                                                "textTransform": "uppercase",
                                                              },
                                                              Object {
                                                                "color": "#fff",
                                                              },
                                                            ]
                                                          }
                                                        >
                                                          Retry
                                                        </Text>
                                                      </Text>
                                                    </View>
                                                  </View>
                                                </View>
                                              </View>
                                            </AnimatedComponent>
                                          </TouchableOpacity>
                                        </Button>
                                      </ThemedComponent>
                                    </Button>
                                  </View>
                                </View>
                              </ThemedComponent>
                            </View>
                          </View>
                        </View>
                      </ThemedComponent>
                    </View>
                  </ComponentState>
                </ThemedComponent>
              </ComponentState>
            </LoadingState>
          </SystemApp>
        </ThemedComponent>
      </SystemApp>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`LoadingState should render timeout message 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <SystemApp>
        <ThemedComponent>
          <SystemApp>
            <LoadingState
              timedOut={true}
            >
              <ComponentState
                description="This is taking longer than usual"
                image={<ActivityIndicator />}
              >
                <ThemedComponent
                  description="This is taking longer than usual"
                  image={<ActivityIndicator />}
                >
                  <ComponentState
                    description="This is taking longer than usual"
                    image={<ActivityIndicator />}
                    styles={
                      Object {
                        "actionRoot": Object {
                          "marginTop": 8,
                        },
                        "description": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                        "image": Object {
                          "height": 250,
                          "width": 250,
                        },
                        "imageRoot": Object {
                          "marginBottom": 16,
                        },
                        "root": Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        },
                        "title": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        }
                      }
                    >
                      <ThemedComponent
                        style={
                          Object {
                            "alignItems": "center",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "justifyContent": "center",
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                                "justifyContent": "center",
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginBottom": 16,
                                }
                              }
                            >
                              <ThemedComponent
                                style={
                                  Object {
                                    "marginBottom": 16,
                                  }
                                }
                              >
                                <View
                                  style={
                                    Object {
                                      "marginBottom": 16,
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "marginBottom": 16,
                                      }
                                    }
                                  >
                                    <ActivityIndicator>
                                      <ThemedComponent>
                                        <ActivityIndicator
                                          animating={true}
                                          color="#999999"
                                          hidesWhenStopped={true}
                                          size="small"
                                        >
                                          <ActivityIndicator
                                            animating={true}
                                            color="#999999"
                                            hidesWhenStopped={true}
                                            size="small"
                                          />
                                        </ActivityIndicator>
                                      </ThemedComponent>
                                    </ActivityIndicator>
                                  </View>
                                </View>
                              </ThemedComponent>
                            </View>
                            <Body2
                              style={
                                Object {
                                  "marginBottom": 8,
                                  "textAlign": "center",
                                }
                              }
                            >
                              <ThemedComponent
                                style={
                                  Object {
                                    "marginBottom": 8,
                                    "textAlign": "center",
                                  }
                                }
                              >
                                <Component
                                  style={
                                    Object {
                                      "marginBottom": 8,
                                      "textAlign": "center",
                                    }
                                  }
                                  styles={
                                    Object {
                                      "root": Object {
                                        "color": "rgba(0, 0, 0, 0.87)",
                                        "fontFamily": "Roboto",
                                        "fontSize": 14,
                                        "fontWeight": "400",
                                      },
                                    }
                                  }
                                >
                                  <Text
                                    style={
                                      Array [
                                        Object {
                                          "color": "rgba(0, 0, 0, 0.87)",
                                          "fontFamily": "Roboto",
                                          "fontSize": 14,
                                          "fontWeight": "400",
                                        },
                                        Object {
                                          "marginBottom": 8,
                                          "textAlign": "center",
                                        },
                                      ]
                                    }
                                  >
                                    <Text
                                      style={
                                        Array [
                                          Object {
                                            "color": "rgba(0, 0, 0, 0.87)",
                                            "fontFamily": "Roboto",
                                            "fontSize": 14,
                                            "fontWeight": "400",
                                          },
                                          Object {
                                            "marginBottom": 8,
                                            "textAlign": "center",
                                          },
                                        ]
                                      }
                                    >
                                      This is taking longer than usual
                                    </Text>
                                  </Text>
                                </Component>
                              </ThemedComponent>
                            </Body2>
                          </View>
                        </View>
                      </ThemedComponent>
                    </View>
                  </ComponentState>
                </ThemedComponent>
              </ComponentState>
            </LoadingState>
          </SystemApp>
        </ThemedComponent>
      </SystemApp>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;
